core/src/cluster_info.rs

```diff
@@ -917,11 +917,14 @@ impl ClusterInfo {
         let mut messages = vec![];
         while !msgs.is_empty() {
             let mut payload = vec![];
-            let mut size = serialized_size(&payload).expect("Couldn't check size");
+            let base_size = serialized_size(&payload).expect("Couldn't check size");
+            let mut size = base_size;
             while let Some(msg) = msgs.pop() {
                 let msg_size = msg.size();
+                // If the message is too big to fit in this batch
                 if size + msg_size > MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
-                    if msg_size < MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
+                    // See if it can fit in the next batch
+                    if base_size + msg_size <= MAX_PROTOCOL_PAYLOAD_SIZE as u64 {
                         msgs.push(msg);
```

Comment:
this is the problem.  the code pushes back into the data structure it iterates over

Reply:
but getting rid of the loop is the way to go.

Reply:
@aeyakovenko how about now?